### DISCUSSION TITLE
Is it possible to create both entities of a bidirectional relationships in the same transaction?

### DISCUSSION THREAD

**Discussion by leonardbinet** [mikro-orm/mikro-orm#6266] [Q&A] [answered]
I have some cases where a bidirectional relationship would be handful.

Let's take the example of expressing a dimension (Mass, Weight, Volume etc), and its units (for Mass: kg, ton, g etc): 
- each unit references a dimension
- and it would be handful in my case to declare a "reference unit" for each dimension

```

@Entity()
class Dimension {

  @PrimaryKey({type: 'varchar'})
  id: string;

  @Property()
  name: string;

  @OneToOne(() => Unit, {deferMode: DeferMode.INITIALLY_DEFERRED})
  referenceUnit: Ref<Unit>;

  constructor(id: string, name: string, referenceUnit: Ref<Unit>) {
    this.id = id;
    this.name = name;
    this.referenceUnit = referenceUnit;
  }
}


@Entity()
class Unit {

  @PrimaryKey({type: 'varchar'})
  id: string;

  @Property()
  name: string;

  @ManyToOne(() => Dimension, {deferMode: DeferMode.INITIALLY_DEFERRED})
  dimension: Ref<Dimension>;

  @Property({type: 'float8'})
  conversionFactor: number;

  constructor(id: string, name: string, dimension: Ref<Dimension>, conversionFactor: number) {
    this.id = id;
    this.name = name;
    this.dimension = dimension;
    this.conversionFactor = conversionFactor;
  }
}

```

yet doing so:
```

test('basic CRUD example', async () => {
  const dimension = orm.em.create(Dimension, { id: 'MASS', name: 'Mass', referenceUnit: ref(Unit, 'KILOGRAM') });
  orm.em.create(Unit, { id: 'KILOGRAM', name: 'Kilogram', dimension: ref(dimension), conversionFactor: 1 });
  orm.em.create(Unit, { id: 'TON', name: 'Ton', dimension: ref(dimension), conversionFactor: 1000 });
  orm.em.create(Unit, { id: 'GRAM', name: 'Gram', dimension: ref(dimension), conversionFactor: 0.001 });
  await orm.em.flush();
  orm.em.clear();

  const units = await orm.em.findAll(Unit);
  expect(units).toHaveLength(3);

});
```

fails 
<img width="1607" alt="Screenshot 2024-12-03 at 10 37 15" src="https://github.com/user-attachments/assets/091afc4a-4906-456e-b526-4d9ea41b6af7">

cf https://github.com/leonardbinet/mikroorm-reproduction/pull/3

**Comment by B4nan** [comment]
Yes, this is possible. Complete repre welcome, either a problem in your entity definition or a bug on our end. My guess its about the reference wrapper, since our tests for this are not using it.

**Reply by leonardbinet** [reply]
here is the repro: https://github.com/leonardbinet/mikroorm-reproduction/pull/3

**Reply by B4nan** [reply]
Yes, it's about the reference wrapper support in `driver.nativeInsertMany`.

**Reply by B4nan** [reply, marked as answer]
The problem is actually elsewhere, the first `em.create` call creates the entity like this:

    Dimension {
      name: 'Mass',
      referenceUnit: (Unit) { id: Ref<Unit<KILOGRAM>> { entity: [(Unit)] } },
      id: 'MASS'
    }

So the `referenceUnit` gets created wrongly, you can see the `Unit.id` is the reference itself. But even if I rework the way entities are created to not use the unmanaged reference, it fails, since internally, we try to make this work regardless of the deferred constraints blindly by removing the cycle and trying to run a separate update query to create it (which obviously fails when the column is not nullable). 

Btw you are missing `ref: true` in the repro, but I guess you are using tsmorph in your app so you forgot to use that explicitly here. Also, `em.create` is there so you don't have to care about the reference wrapper, you can just pass the FK value there and it will handle the rest for you, no need for doing `dimension: ref(dimension)`. You should be using the `ref` helper inside the constructor if something, that was the main idea behind adding it, so your constructor can support `unit: string | Unit | Ref<Unit>` and you can do `ref(Unit, unit)` (or `rel(Unit, unit)`) to normalize it.

I've split this into two fixes: db7cb8ab350663c0e4db0640aee3dd8c24a5acfa and 85a649b04059f83ab3f3e2de6da6134801877bb3.

**Reply by leonardbinet** [reply]
thanks a lot 🙏

**Reply by leonardbinet** [reply]
re-reading this comment `Btw you are missing ref: true in the repro, but I guess you are using tsmorph in your app so you forgot to use that explicitly here` I was wondering what is the impact of not adding `ref: true` (given I'm indeed using tsmorph), does it have any effect?

**Reply by B4nan** [reply]
None, it gets inferred with tsmoprh. But in your test you were using reflect metadata provider, which can't infer this, so you'd end up with the ORM hydrating it without the `Reference` wrapper.

**Comment by leonardbinet** [comment]
👋 , I have the same question, for entities with auto-incremented ids, how can I perform the same as `tests/issues/GH6266.test.ts`, but with id columns being auto-incremented integers?

what would be the right manner to create my entities?
<img width="1125" height="402" alt="Screenshot 2025-09-18 at 14 23 27" src="https://github.com/user-attachments/assets/f461cc92-5328-422d-a02d-e15e6a2eb0d9" />
<img width="740" height="308" alt="Screenshot 2025-09-18 at 14 23 30" src="https://github.com/user-attachments/assets/307b9ea7-a63a-46a5-b98b-7051d4f40cf2" />

**Reply by leonardbinet** [reply]
Here is a try, translated as a test: https://github.com/mikro-orm/mikro-orm/pull/6859

The weird thing is that this worked in our codebase until v6.4.15 (very similar code, but obviously I failed to reproduce since when I run this test on v6.4.15, the test fails), and breaks while bumping to v6.5.4.

I would be interested in knowing if this is supposed to be supported, and if so how it should be used.


Note, the fact that:
- the `dimension` entity has a `referenceUnit`
- but the `loadedDimension` doesn't (after fetch from db)
looks like a bug

**Reply by leonardbinet** [reply]
@B4nan, given that discussion is resolved, would you prefer me to open a new dedicated discussion for this topic, or should I leave it here?

**Reply by B4nan** [reply]
Given you opened the PR with the test case, I guess I don't need a new one, it's enough of a reminder. Will take a look soon, but need to resolve some things I got in progress already.

**Reply by leonardbinet** [reply]
note for anyone checking this issue, resolved cf https://github.com/mikro-orm/mikro-orm/pull/6859#issuecomment-3314817997
